### DISCUSSION TITLE
How to customize the theme more elegantly?

### DISCUSSION THREAD

**Discussion by YunYouJun** [element-plus/element-plus#3661] [Ideas]
We found that `theme-chalk` has many remaining problems with inconsistent naming mechanisms.
We will refactor the variables in theme-chalk to a certain extent, and finally provide an example of a customized theme.

The break modification will be recorded in the discussion.

If you have any suggestions, please leave a message here.

vite & custom theme examples:

- On demand with unplugin-vue-components: [element-plus-vite-starter](https://github.com/element-plus/element-plus-vite-starter)
- On demand manually with unplugin-element-plus: [unplugin-element-plus/examples/vite](https://github.com/element-plus/unplugin-element-plus)

---

Variables start with `_` & `-` will be considered private variables. They can not be custom from outside.

```scss
$--colors-primary: blue !default; // can not be custom from outside
$color-primary: blue !default; // can be custom from outside
```

> https://sass-lang.com/documentation/at-rules/import

We have to rename all varibales start with `$--` to `$`.

---

- `-background-color` -> `-bg-color`
- `-font-color` -> `-text-color`

Example:
- `$--button-font-color` will change to `$--button-text-color` -> `$button-text-color`
- `$--button-background-color` will change to `$--button-bg-color` -> `$button-bg-color`

> Similiar to tailwindcss

**Comment by xkubow** [comment]
I'm not very good with css. I'm struggling to modify padding on .el-main. Could you provide some examples for modifying the .el-main?

I did try:
```
.el-main {
--el-main-padding:1px
}
```

```
// main
$main: () !default;
$--main: map.merge(
  (
    'padding': 0px,
  ),
  $main
);
```

**Reply by YunYouJun** [reply]
You can just write:

> Override see https://element-plus.org/en-US/guide/theming.html#how-to-override-it

```css
// styles/element/index.scss
/* just override what you need */
@forward 'element-plus/theme-chalk/src/common/var.scss' with (
  $main: (
    'padding': 1px,
  )
);
```

```ts
import Vue from 'vue'

import './styles/element/index.scss'
import ElementPlus from 'element-plus'
import App from './App.vue'

const app = createApp(App)
app.use(ElementPlus)
```

**Reply by YunYouJun** [reply]
If you just want to custom a `<el-main>`, you can like this:

```vue
<template>
  <el-main style="--el-main-padding: 10px;"></el-main>
</template>
```

**Reply by orenmizr** [reply]
@YunYouJun is there a repo in the real world where someone has successfully changed the the theme across multiple components ?

**Reply by YunYouJun** [reply]
Sorry, I'm a little confused.
Do you mean changing the content of the theme as a whole (not dark mode) such as the color scheme?

**Comment by mp5maker** [comment]
@YunYouJun 

I would like to thank you for maintaining this. 
Unfortunately, I having difficulty going through theming section

**Issues**
I still could not figure how to switch between custom themes

Let just say I have ['light', 'dark', 'dark-blue', 'light-blue']
dark-blue is a variant of a dark theme
light-blue is a variant of a light theme

I want to keep to keep the configurations in four different files
and load them accordingly. I also want to load theme `on-demand`

Are there any examples for this ?

**Reply by orenmizr** [reply]
if you succeeded in building a theme. please publish the setup. i also had issues - css vars vs scss. tried to modify a component completely -  i think it was the table. could not style it to my liking.

**Reply by YunYouJun** [reply]
Hello, what does loading on demand mean?

Do you want to switch only at the code level, or during the runtime of your website?

**Reply by YunYouJun** [reply]
Maybe we can open a `feature request` issue to https://github.com/element-plus/element-plus-vite-starter .

Just `@` me there, I wll write an example for you.

**Comment by gimjin** [comment]
JavaScript framework and UI component library saved a lot of company money. Now the requirements for front-end engineers are extremely low for css.

So maybe something like naive-ui having a UI that automatically generates css/scss might be elegant. Writing css/scss directly is advanced.

Tell you something interesting @YunYouJun , I saw a senior front-end engineer abandoning element-plus and using naive-ui in order to achieve theme switching.

**Reply by YunYouJun** [reply]
`naive-ui` is a great library, and while we seem to be in competition, the author and team members of naive-ui are actually my friends.

From the beginning, however, naive-ui's implementation strategy in terms of style was different from element's. naive-ui is similar to the CSS-in-js scheme, which has a lot of flexibility at the expense of runtime consumption and readability, while element-plus inherits from element-ui's scss compilation scheme, which follows the BFC specification. but also sacrifices some flexibility.

This is actually the current controversy, CSS-IN-JS VS Style Preprocessor. You can find many articles on the web judging the advantages and disadvantages, and I don't want to give a biased opinion on this. But I guess it's because different solutions are explored in different ways that the existence of different component libraries becomes more meaningful, isn't it?
